### PR TITLE
[FLINK-31432] Introduce a special StoreWriteOperator to deal with schema changes

### PR DESCRIPTION
**Pull request by tsreaper**
### Purpose

Currently `StoreWriteOperator` is not able to deal with schema changes. We need to introduce a special `StoreWriteOperator` to deal with schema changes.

### Tests

* SchemaAwareStoreWriteOperatorTest

### API and Format 

N/A

### Documentation

N/A

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/cdc/Record.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.paimon.cdc;
+
+import org.apache.paimon.types.RowKind;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.Objects;
+
+/** A data change message from the CDC source. */
+public class Record implements Serializable {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/cdc/Record.java:28]: Maybe `RichRecord`? You need something to indicate `Map<String, String> fields` with field names.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/SchemaAwareStoreWriteOperator.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+/**
+ * An {@link AbstractStoreWriteOperator} which is aware of schema changes.
+ *
+ * <p>When the input {@link Record} contains a field name not in the current {@link TableSchema}, it
+ * periodically queries the latest schema, until the latest schema contains that field name.
+ */
+public class SchemaAwareStoreWriteOperator extends AbstractStoreWriteOperator<Record> {
+
+    static final ConfigOption<Duration> SCHEMA_UPDATE_WAIT_TIME =
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/SchemaAwareStoreWriteOperator.java:48]: retry sleep time?
It's easy to misunderstand it as just waiting for 500 ms.

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/SchemaAwareStoreWriteOperator.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        TableSchema schema = table.schema();
+        GenericRow row = new GenericRow(schema.fields().size());
+        row.setRowKind(record.kind());
+        for (Map.Entry<String, String> field : fields.entrySet()) {
+            String key = field.getKey();
+            String value = field.getValue();
+            int idx = schema.fieldNames().indexOf(key);
+            DataType type = schema.fields().get(idx).type();
+            row.setField(idx, TypeUtils.castFromString(value, type));
```
Comment (JingsongLi) [inline paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/cdc/SchemaAwareStoreWriteOperator.java:87]: `TypeUtils.castFromString` is limited, can not support complex types.
Add a TODO here?

File: paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreSinkWriteImpl.java [hunk 4]
```diff
@@ -136,4 +135,16 @@ public void close() throws Exception {
             write.close();
         }
     }
+
+    @Override
+    public void replace(Function<String, TableWriteImpl<?>> newWriteProvider) throws Exception {
+        if (commitUser == null) {
+            return;
```
[the remark this answers is not in the document]
Comment (tsreaper) [reply paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/sink/StoreSinkWriteImpl.java:142]: See line 64~69.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**